pkg/registry/apis/datasource/register.go

```diff
@@ -213,6 +215,7 @@ func (b *DataSourceAPIBuilder) GetAPIGroupInfo(
 	}
 	storage[conn.StoragePath("query")] = &subQueryREST{builder: b}
 	storage[conn.StoragePath("health")] = &subHealthREST{builder: b}
+	storage[conn.StoragePath("query-convert")] = &subQueryConvertREST{builder: b}
```

Comment:
do we want them to be in the context of a datasource instance?  EG, or should we try to expose it one level up?

EG, can we convert a prometheus query (at a given apiversion) without having all the credentials?

Reply:
Likewise -- how do we know which data sources implement convert?  prepare? sanitize?

can we only expose this method when it exists?   If that is true, do we need a feature flag for it?

Reply:
> do we want them to be in the context of a datasource instance? EG, or should we try to expose it one level up?
> EG, can we convert a prometheus query (at a given apiversion) without having all the credentials?

I think it's more feature complete if it's at this level. For example, to migrate a prometheus query to a new version, it may require to hit an external API, therefore it may need credentials.

> Likewise -- how do we know which data sources implement convert? prepare? sanitize?
> can we only expose this method when it exists? If that is true, do we need a feature flag for it?

We don't know here if it exists or not (it always exists from this point of view, it just returns an "unimplemented" error if it doesn't). It's the same than the `resource` endpoint. It's up to the client to decide if it should call it or not.

Reply:
I get that we do not know, but can we add something so we do know?  (maybe not in this PR?)  It would be really nice to only expose things that exist so they only show up in the openapi spec when they exist.  (like for proxy etc)

Reply:
right, that probably needs its own discussion / implementation so we include more things like resources the plugin expose or other optional endpoints.

Reply:
> For example, to migrate a prometheus query to a new version, it may require to hit an external API, therefore it may need credentials.

Maybe a bad example using Prometheus where I don't know how likely it is with this requirement, but maybe it is? With the (now old) thema solution we did only support migration from A to B and to support more advanced logic, such as database lookups, you had to use a custom lacuna (or whatever we called it). Based on this and to simplify I would suggest "expose it one level up" to support the simplest use case. 

> right, that probably needs its own discussion / implementation so we include more things like resources the plugin expose or other optional endpoints.

Yes, sounds like something we've discussed for some time and having an idea about without really investigating what our options might be. Lets create a design doc to discuss further?

Reply:
> Maybe a bad example using Prometheus where I don't know how likely it is with this requirement, but maybe it is? With the (now old) thema solution we did only support migration from A to B and to support more advanced logic, such as database lookups, you had to use a custom lacuna (or whatever we called it). Based on this and to simplify I would suggest "expose it one level up" to support the simplest use case.
> 

I don't know if it would be any simpler but maybe I am missing things. @ryantxu could you elaborate how this may look like if moved to a "level up"?

Reply:
a level up would be without plugin settings -- essentially can we validate/mutate the 

rather than:
http://localhost:3000/apis/testdata.datasource.grafana.app/v0alpha1/namespaces/default/connections/PD8C576611E62080A/convert

http://localhost:3000/apis/testdata.datasource.grafana.app/v0alpha1/namespaces/default/convert

(note there is no settings reference -- in the plugin we could send an empty one or pick any) but importantly get code that can run without live settings so we can bulk fix things easily in the future